lib/data_migrate/data_migrator.rb

```diff
@@ -81,4 +97,90 @@ def db_config
 
     end
   end
+
+  ##
+  # This class extends DatabaseTasks to override the schema_file method.
+  class DatabaseTasks
+    extend ActiveRecord::Tasks::DatabaseTasks
+
+    def self.schema_file(format = ActiveRecord::Base.schema_format)
+      case format
+      when :ruby
+        File.join(db_dir, "data_schema.rb")
+      else
+        message = "Only Ruby-based data_schema files are supported at this time."
+        Kernel.abort message
+      end
+    end
+  end
+
+  ##
+  # Provides the definition method for data_schema.rb
+  class Data < ActiveRecord::Schema
+    # This method is based on the following two methods
+    #   ActiveRecord::Schema#define
+    #   ActiveRecord::ConnectionAdapters::SchemaStatements#ssume_migrated_upto_version
+    def define(info)
+      DataMigrate::DataMigrator.assure_data_schema_table
+
+      return if info[:version].blank?
+
+      version = info[:version].to_i
+      sm_table = quote_table_name(DataMigrate::DataMigrator.schema_migrations_table_name)
+      migrated = select_values("SELECT version FROM #{sm_table}").map(&:to_i)
+
+      versions = []
+      Dir.foreach(DataMigrate::DataMigrator.full_migrations_path) do |file|
+        match_data = DataMigrate::DataMigrator.match(file)
+        versions << match_data[1].to_i if match_data
+      end
+
+      unless migrated.include?(version)
+        execute "INSERT INTO #{sm_table} (version) VALUES ('#{version}')"
+      end
+
+      inserted = Set.new
+      (versions - migrated).each do |v|
+        if inserted.include?(v)
+          raise "Duplicate data migration #{v}. Please renumber your data migrations to resolve the conflict."
+        elsif v < version
+          execute "INSERT INTO #{sm_table} (version) VALUES ('#{v}')"
+          inserted << v
+        end
+      end
+    end
+  end
+
+  ##
+  # Provides the capability to write the current data schema version to the data_schema file
+  # Based on ActiveRecord::SchemaDumper
+  class SchemaDumper
```

Comment:
same here, separate file please just to keep things clean (i've been trying to split large files into smaller ones. as well as